core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/io/ContentLengthAwareSubscriber.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.internal.signer.io;
+
+import java.nio.ByteBuffer;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * Decorator subscriber that limits the number of bytes sent to the wrapped subscriber to at most {@code contentLength}. Once
+ * the given content length is reached, the upstream subscription is cancelled, and the wrapped subscriber is completed.
+ */
+@SdkInternalApi
+public class ContentLengthAwareSubscriber implements Subscriber<ByteBuffer> {
```

Comment:
Should we move this class to utils since this may be used elsewhere?

Reply:
Sure makes sense